packages/salesforce-adapter/src/filters/field_references.ts

```diff
@@ -176,39 +175,29 @@ export const addReferences = async (
 ): Promise<void> => {
   const resolverFinder = generateReferenceResolverFinder(defs)
 
-  const elementsWithFields = flatMapAsync(
+  const elementsAndFields = flatMapAsync(
     await referenceElements.getAll(),
     extractFlatCustomObjectFields,
   )
-  // TODO - when transformValues becomes async the first index can be to elemID and not the whole
-  // element and we can use the element source directly instead of creating the second index
-  const { elemLookup, elemByElemID } = await multiIndex
-    .buildMultiIndex<Element>()
-    .addIndex({
-      name: 'elemLookup',
-      filter: hasApiName,
-      key: async (elem) => [await metadataType(elem), await apiName(elem)],
-    })
-    .addIndex({
-      name: 'elemByElemID',
-      filter: (elem) => !isField(elem),
-      key: (elem) => [elem.elemID.getFullName()],
-    })
-    .process(elementsWithFields)
+  const elemIDLookup = await multiIndex.keyByAsync({
+    iter: elementsAndFields,
+    filter: hasApiName,
+    key: async (elem) => [await metadataType(elem), await apiName(elem)],
+    map: (elem) => elem.elemID,
+  })
 
   const fieldsWithResolvedReferences = new Set<string>()
-  await awu(elements)
-    .filter(isInstanceElement)
-    .forEach(async (instance) => {
-      instance.value = await replaceReferenceValues({
-        instance,
-        resolverFinder,
-        elemLookupMaps: { elemLookup },
-        fieldsWithResolvedReferences,
-        elemByElemID,
-        contextStrategyLookup,
-      })
+  const instances = elements.filter(isInstanceElement)
+  await awu(instances).forEach(async (instance) => {
+    instance.value = await replaceReferenceValues({
+      instance,
+      resolverFinder,
+      elemIDLookupMaps: { elemIDLookup },
+      fieldsWithResolvedReferences,
+      elementsSource: referenceElements,
+      contextStrategyLookup,
     })
+  })
```

Comment:
would this be faster if we let it run on more instances concurrently?

```suggestion
  await withLimitedConcurrency(
    elements
      .filter(isInstanceElement)
      .map(instance => async () => {
        instance.value = await replaceReferenceValues({
          instance,
          resolverFinder,
          elemIDLookupMaps: { elemIDLookup },
          fieldsWithResolvedReferences,
          elementsSource: referenceElements,
          contextStrategyLookup,
        })
      }),
    20 // any reasonable number that is larger than 1 but not too much to avoid memory issues, and of course it should be a const and not literally 20 here
  )
```

Reply:
Only one way to find out.

Reply:
From a single (large) test this actually seems to be slower so I'll go back to the first implementation.